File: Live Class - Java/StockPriceManager.java


import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.locks.ReentrantLock;

class StockPrice {
    private final double price;
    private final long timestamp; // Timestamp in milliseconds

    public StockPrice(double price, long timestamp) {
        this.price = price;
        this.timestamp = timestamp;
    }

    public double getPrice() {
        return price;
    }

    public long getTimestamp() {
        return timestamp;
    }
}

public class StockPriceManager {
    private final ConcurrentHashMap<String, StockPrice> stockPrices = new ConcurrentHashMap<>(); // Registry for stock symbols
    private final ConcurrentHashMap<String, ReentrantLock> locks = new ConcurrentHashMap<>(); // Locks for each symbol

    // Method to get the lock for a specific stock symbol
    private ReentrantLock getLock(String symbol) {
        return locks.computeIfAbsent(symbol, k -> new ReentrantLock());
    }

    // Method to update the stock price for a symbol if the new price is more recent
    public void updateStockPrice(String symbol, double newPrice, long timestamp) {
        ReentrantLock lock = getLock(symbol); // Get lock for the specific symbol
        lock.lock(); // Acquire the lock
        try {
            StockPrice currentPrice = stockPrices.get(symbol);
            if (currentPrice == null || timestamp > currentPrice.getTimestamp()) {
                stockPrices.put(symbol, new StockPrice(newPrice, timestamp));
                System.out.println("Stock price for " + symbol + " updated to: " + newPrice + " at timestamp: " + timestamp);
            } else {
                System.out.println("Ignored outdated stock price update for " + symbol + ": " + newPrice + " at timestamp: " + timestamp);
            }
        } finally {
            lock.unlock(); // Release the lock
        }
    }

    // Method to get the current stock price for a symbol
    public StockPrice getStockPrice(String symbol) {
        ReentrantLock lock = getLock(symbol); // Get lock for the specific symbol
        lock.lock(); // Acquire the lock
        try {
            return stockPrices.get(symbol);
        } finally {
            lock.unlock(); // Release the lock
        }
    }

    public static void main(String[] args) {
        StockPriceManager manager = new StockPriceManager();

        // Simulate updating and retrieving stock prices in concurrent threads
        Runnable updateTask = () -> {
            for (int i = 0; i < 5; i++) {
                String symbol = "AAPL"; // Example stock symbol
                manager.updateStockPrice(symbol, Math.random() * 1000, System.currentTimeMillis()); // Update with random price and current timestamp
                try {
                    Thread.sleep(100); // Simulate time delay
                } catch (InterruptedException e) {
                    Thread.currentThread().interrupt();
                }
            }
        };

        Runnable retrieveTask = () -> {
            for (int i = 0; i < 10; i++) {
                String symbol = "AAPL"; // Example stock symbol
                StockPrice stockPrice = manager.getStockPrice(symbol);
                if (stockPrice != null) {
                    System.out.println("Current stock price for " + symbol + ": " + stockPrice.getPrice() + " at timestamp: " + stockPrice.getTimestamp());
                } else {
                    System.out.println("No stock price available for " + symbol);
                }
                try {
                    Thread.sleep(200); // Simulate time delay
                } catch (InterruptedException e) {
                    Thread.currentThread().interrupt();
                }
            }
        };

        // Create and start threads
        Thread updaterThread = new Thread(updateTask, "UpdaterThread");
        Thread retrieverThread = new Thread(retrieveTask, "RetrieverThread");

        updaterThread.start();
        retrieverThread.start();

        try {
            updaterThread.join();
            retrieverThread.join();
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
    }
}

File: Live Class - Python/stock_price_manager.py
import threading
import time
from collections import defaultdict

class StockPrice:
    def __init__(self, price, timestamp):
        self.price = price
        self.timestamp = timestamp

    def __repr__(self):
        return f"StockPrice(price={self.price}, timestamp={self.timestamp})"


class StockPriceManager:
    def __init__(self):
        self.stock_prices = {}
        self.locks = defaultdict(threading.Lock)  # Lock for each symbol

    def update_stock_price(self, symbol, new_price, timestamp):
        lock = self.locks[symbol]
        with lock:  # Acquire the lock for the specific symbol
            current_price = self.stock_prices.get(symbol)
            if current_price is None or timestamp > current_price.timestamp:
                self.stock_prices[symbol] = StockPrice(new_price, timestamp)
                print(f"Stock price for {symbol} updated to: {new_price} at timestamp: {timestamp}")
            else:
                print(f"Ignored outdated stock price update for {symbol}: {new_price} at timestamp: {timestamp}")

    def get_stock_price(self, symbol):
        lock = self.locks[symbol]
        with lock:  # Acquire the lock for the specific symbol
            return self.stock_prices.get(symbol)

def update_task(manager):
    for _ in range(5):
        symbol = "AAPL"  # Example stock symbol
        manager.update_stock_price(symbol, random.random() * 1000, int(time.time() * 1000))  # Update with random price and current timestamp
        time.sleep(0.1)  # Simulate time delay

def retrieve_task(manager):
    for _ in range(10):
        symbol = "AAPL"  # Example stock symbol
        stock_price = manager.get_stock_price(symbol)
        if stock_price:
            print(f"Current stock price for {symbol}: {stock_price.price} at timestamp: {stock_price.timestamp}")
        else:
            print(f"No stock price available for {symbol}")
        time.sleep(0.2)  # Simulate time delay

if __name__ == "__main__":
    import random

    manager = StockPriceManager()

    # Create and start threads
    updater_thread = threading.Thread(target=update_task, args=(manager,), name="UpdaterThread")
    retriever_thread = threading.Thread(target=retrieve_task, args=(manager,), name="RetrieverThread")

    updater_thread.start()
    retriever_thread.start()

    updater_thread.join()
    retriever_thread.join()
